### PR TITLE
Expose `ProcessConfig` and friends from the internal module

### LINKED ISSUE #62: Expose structure of `ProcessConfig` from internal module.
I wanted to write a wrapper of `runProcess` which would print out the command which was going to run before running it. However, the only way I have display a `ProcessConfig` is using the `Show` instance which prints out a lot more information (including full environment). 

Ideally it would be good to expose the structure of `ProcessConfig` via the internal module (so I can write the function myself).

### PR DESCRIPTION
**Pull request by mmhat**
I basically moved all the stuff concerned with `ProcessConfig` (e.g. setters, smart constructors, ...) to the internal module and re-export it in the public API. This structure appeared simpler to me than having some of the bindings in the internal module and some in the public one: There are a bunch of top-level bindings (like `shell`, `proc`) needed in the instances of the datatypes and unless we go with orphan instances in `System.Process.Typed` those need to live in the internal module.

Fixes #62

### CODE REVIEW

File: package.yaml [hunk 1]
```diff
@@ -32,6 +32,8 @@ tests:
   typed-process-test:
     main: Spec.hs
     source-dirs: test
+    build-tools:
+    - hspec-discover:hspec-discover >=2 && <3
```
Comment (tomjaguarpaw) [inline package.yaml:36]: I don't understand why this is needed.

Comment (mmhat) [reply package.yaml:36]: I added in the case one wants to run the tests with cabal:
```shell
hpack
cabal test all
```
Without the generated `build-tools-depends` entry in the Cabal file you get a `could not execute: hspec-discover`. For stack it is indeed not necessary.

**Comment by tomjaguarpaw** [comment]
These three commits seems logically independent to me, so can you please make them in three separate PRs? Having them in the same PR makes it harder to review.

**Comment by mmhat** [comment]
@tomjaguarpaw I moved the unrelated changes to https://github.com/fpco/typed-process/pull/64 and https://github.com/fpco/typed-process/pull/65.

**Comment by tomjaguarpaw** [comment]
Thanks.

File: src/System/Process/Typed/Internal.hs [hunk 2]
```diff
@@ -1,7 +1,702 @@
...
+
+      -- * Exceptions
+    , ExitCodeException (..)
+    , ByteStringOutputException (..)
+
+      -- * Helper functions
+    , nullDevice
+    , bracket
+    , finally
+    ) where
```
Comment (tomjaguarpaw) [inline src/System/Process/Typed/Internal.hs:74]: Is there any point restricting the exports for the `Internal` module? As an `Internal` module I would expect it to export everything.

Comment (mmhat) [reply src/System/Process/Typed/Internal.hs:74]: No, there is no particular reason. It is simply a leftover from the refactoring; I'll remove the export list.

Comment (mmhat) [reply src/System/Process/Typed/Internal.hs]: Done.

**Comment by tomjaguarpaw** [comment]
Not sure why the MacOS tests failed. I guess it's a problem with the tests, not this PR.

**Comment by tomjaguarpaw** [comment]
Thanks for this, and thanks for being responsive to requests!

**Comment by mmhat** [comment]
@tomjaguarpaw Thank you for the quick review!

**Pull request merged**